src/compiler/checker.ts

```diff
@@ -45863,6 +45863,9 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
 
         let allIterationTypes: IterationTypes[] | undefined;
         for (const constituent of (type as UnionType).types) {
+            if (!!(getReducedType(constituent).flags & TypeFlags.Never)) {

```

Comment:
q: what if all union members reduce to never? wouldn't this break a case like this?
```ts
declare const o2: ({ a: "foo" } & { a: "bar" }) | ({ b: "qwe" } & { b: "rty" });
const [el2] = o2; // error
```